refactor(common): Use safe interfaces for `DataChunkBuilder`

kwannoel opened the pull request:
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Sqlsmith found a bug in arrangement backfill where the same row is read twice in snapshot. That's because initial buffer is not flushed.

To be defensive, the same is done for no_shuffle_backfill, although it's not strictly necessary.

Additionally, we also refactor some interfaces for `DataChunkBuilder`, making sure they are safe.

<!--

**Please do not leave this empty!**

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)
- Refer to a related PR or issue link (optional)

-->

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)
<!-- To manually trigger a benchmark, please check out [Notion](https://www.notion.so/risingwave-labs/Manually-trigger-nexmark-performance-dashboard-test-b784f1eae1cf48889b2645d020b6b7d3). -->
- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.


<!--
Please create a release note for your changes.

Discuss technical details in the "What's changed" section, and
focus on the impact on users in the release note.

You should also mention the environment or conditions where the impact may occur.
-->

</details>

BugenZhao commented:
When reviewing this PR, I found several issues with the new `DataChunkBuilder` design. Would you please first take a look? Sorry for being late. 🥺

https://github.com/risingwavelabs/risingwave/pull/11117#pullrequestreview-1606174529

kwannoel commented:
> When reviewing this PR, I found several issues with the new `DataChunkBuilder` design. Would you please first take a look? Sorry for being late. 🥺
> 
> [#11117 (review)](https://github.com/risingwavelabs/risingwave/pull/11117#pullrequestreview-1606174529)

Addressed the review comments, thanks for the feedback.

chenzl25 reviewed (commented):
May I ask how to construct such a case that can hit this bug?

kwannoel commented on src/stream/src/executor/backfill/no_shuffle_backfill.rs:
```diff
@@ -163,16 +163,20 @@ where
+                };
+                let snapshot_buffer_is_empty = builder.is_empty();
+                builder.clear();
+                snapshot_is_empty && snapshot_buffer_is_empty
```
Here's the code of interest cc @chenzl25 .

To reproduce it:
1. Initial snapshot read occurs, but does not produce a chunk. Instead it only fills up the builder's buffer. Let's say row A is now in this buffer.
2. When backfill occurs later, snapshot is refreshed, it will read row A again.
3. Now builder's buffer has 2x row A, if it was not cleared after 1.

The test was originally found in arrangement backfill.
I think the same issue could be encountered in `no_shuffle_backfill` as well, but don't have a test case for it.

chenzl25 replied:
However, I can't construct such a case manually. Actually, I want to evaluate the scope of impact.

chenzl25 replied:
Since ` snapshot.try_next().await?.unwrap().is_none()` will wait until a chunk returns. The builder's buffer should contain no extra row I think.

kwannoel replied:
OH it must be arrangement backfill specific.
You raise a good point. I think it only happens in arrangement backfill, and not here, because snapshot_read per vnode returns a stream of chunks.

To be precise, the following scenario occurs:
1. Arrangement backfill does initial snapshot read, this occurs for multiple vnode streams.
2. Perhaps 1 vnode stream returns a chunk, so it continues. Concurrently, another vnode stream is in process of constructing its chunk, but did not finish it.
3. Then what happens is that for the first vnode stream which returned a chunk, it will clean up the corresponding builder.
4. But for the subsequent vnode stream which did not complete it, it will still have rows in the buffer, which need to be cleaned.

This does not apply here, with a single snapshot read stream.

I will remove the "bugfix" in this PR, and just address bugen's comments on API changes here.

kwannoel replied:
Oh but actually could there be an (unlikely) scenario where `iter_chunks` continues executing on row_iter?

i.e. the following sequence of events:
```rs
                    let snapshot = Self::snapshot_read(
                        &self.upstream_table,
                        init_epoch,
                        None,
                        false,
                        &mut builder,
                    );
                    pin_mut!(snapshot);
                    snapshot.try_next().await?.unwrap().is_none() -> first chunk is processed.
  --> here the next snapshot chunk is being processed asynchronously.
} --> drop is called before the chunk is finished processing
```

chenzl25 replied:
If we don't call the `next`, I think it is safe. Anyway, clear the buffer at the end LGTM.

chenzl25 reviewed (approved):
LGTM

BugenZhao commented on src/storage/src/table/mod.rs:
```diff
@@ -130,28 +130,22 @@ where
-    } else {
-        Ok(Some(chunk))
-    }
+    let chunk = builder.consume_all();
+    Ok(chunk)
```
This seems a little bit weird to me. I assume this method will be called in a loop. If we always consume the chunk builder here, then what's the difference compared to initializing a `ChunkBuilder` inside this function?

kwannoel replied:
Actually it was also unintuitive to me. The reference scenario is here: https://github.com/risingwavelabs/risingwave/issues/11007#issuecomment-1681914555.

But idea is this. Suppose snapshot read proceeds midway.

It can get stuck here while waiting for next row, perhaps object store latency:
```
            Some(row) => {
    while let Some(row) = stream.next().await.transpose()? 
...
```

Now there are some rows in the buffer before this, but they are not yet flushed.

If some upstream updates come, we will process those, and if within the upstream updates, there is a barrier, we will drop the snapshot.

But critically even though the snapshot is dropped, there are still rows inside the buffer.

We don't want to just drop these rows, losing the progress we made.

So we should still flush these rows out, and that is done outside of the snapshot_read function, only on barrier.

BugenZhao replied:
Yeah, this makes much sense. In other words, this stream will likely be canceled in its usage cases.

kwannoel commented:
No need to cherry-pick as it's not actually "fixing" anything.
